src/tree_data.c

```diff
@@ -2779,6 +2779,82 @@ lyd_target(const struct ly_path *path, const struct lyd_node *tree)
     return (struct lyd_node_term *)target;
 }
 
+/**
+ * @brief Compare schemas of two Compiled YANG data nodes.
+ *
+ * @param schema1 of first node.
+ * @param schema2 of second node.
+ * @return 0 if the schemas are equivalent otherwise 1.
+ */
+static ly_bool
+lyd_compare_schemas(const struct lysc_node *schema1, const struct lysc_node *schema2)
+{
+    if (schema1 == schema2) {
+        return 0;
+    } else if (!schema1 || !schema2) {
+        return 1;
+    }
+
+    if (schema1->nodetype != schema2->nodetype) {
+        return 1;
+    }
+
+    if (schema1->module->latest_revision != schema2->module->latest_revision) {
+        return 1;
+    }
+
+    if (strcmp(schema1->name, schema2->name)) {
+        return 1;
+    }
+
+    if (strcmp(schema1->module->name, schema2->module->name)) {
+        return 1;
+    }
+
+    if (schema1->module->revision || schema2->module->revision) {
+        if (!schema1->module->revision || !schema2->module->revision) {
+            return 1;
+        }
+        if (strcmp(schema1->module->revision, schema2->module->revision)) {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+/**
+ * @brief Compare all parental schemas of two data nodes.
+ *
+ * @param schema1 of first node.
+ * @param schema2 of second node.
+ * @return 0 if the all parental schemas are equivalent otherwise 1.
+ */
+static ly_bool
+lyd_compare_parental_schemas(const struct lyd_node *node1, const struct lyd_node *node2)
+{
+    const struct lyd_node_inner *parent1, *parent2;
+
+    if (node1->schema == node2->schema) {
```

Comment:
You already did this comparison, is there any purpose to doing it again? You can put an assert, if you to make sure it works the way it should.

Reply:
assert won't help here. The purpose of this condition is optimization. Thanks to it, it does not have to iterate through the `for `cycle below. I assume that if the schemes are equal, all parent nodes must be equal.

Reply:
Right, I have forgotten a comment I wanted to put here, again distinguishing between the same context and a different one. You do not do that at all and only comparing the schemas for equality is not enough, as I have mentioned.